### PR TITLE
Slice entropy regularization (prevent slice collapse)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis
With 32 slices and n_layers=1, there's a risk that many slices become redundant or that a few slices dominate the assignment — the classical \"dead expert\" problem from mixture-of-experts literature. If only 10 of 32 slices are actively used, the model wastes 2/3 of its representational capacity.

Entropy regularization (used in Switch Transformers, VQ-VAE) keeps all slices active by penalizing low-entropy (peaked) assignment distributions. This forces the model to use its full 32-slice capacity for finer-grained spatial decomposition, which should help surface accuracy since surfaces need distinct representations from neighboring volume regions.

## Baseline
- val/loss: 2.2217
- val_in_dist/mae_surf_p: 21.18
- val_ood_cond/mae_surf_p: 20.47
- val_ood_re/mae_surf_p: 30.95
- val_tandem_transfer/mae_surf_p: 41.23

---

## Results (Revision 2 — includes diagnostic + weight sweep)

Three runs total across the two iterations.

### Run summary

| Run | Weight | W&B ID | val/loss | in_dist surf_p | ood_cond surf_p | ood_re surf_p | tandem surf_p |
|---|---|---|---|---|---|---|---|
| Baseline | — | — | 2.2217 | 21.18 Pa | 20.47 Pa | 30.95 Pa | 41.23 Pa |
| Original run | 0.01 | `jff27zf1` | 2.2683 | 22.33 Pa | 20.50 Pa | 31.40 Pa | 41.13 Pa |
| Diagnostic (no penalty) | 0.0 | `pw6d6w5g` | 2.2400 | 21.07 Pa | 20.66 Pa | 30.65 Pa | 41.02 Pa |
| Gentler weight | 0.001 | `ssuh3ghd` | 2.2873 | 22.91 Pa | 20.64 Pa | 31.05 Pa | 41.44 Pa |

All three entropy-penalty runs are within noise of each other, and both weight=0.01 and weight=0.001 are worse than the diagnostic (no penalty). The diagnostic itself is essentially baseline-level (stochastic variation only).

### Slice entropy measurement (diagnostic run, run `pw6d6w5g`)

Measured from the last-block `slice_weights` averaged across nodes, over 500 training steps:

- **Mean entropy: 2.28 nats** (out of max log(32) = 3.47 nats)
- Range: 1.84 – 2.82 nats
- % of maximum: **65.9%**

The slices are NOT collapsing — they're operating at ~66% of maximum entropy, meaning roughly 2/3 of slices are actively being used. The "dead expert" premise does not hold for this model.

### OOD Re overflow is pre-existing

`val_ood_re/vol_loss = 18.9×10⁹` appears in ALL THREE runs including the no-penalty diagnostic. This is a pre-existing overflow in the `exp-noam/slice-entropy-reg` branch baseline (the same as seen in `exp-noam/fix-ood-re-overflow`), not caused by entropy regularization.

### What happened

The hypothesis is false. The Transolver's softmax-normalized slice assignment naturally maintains ~66% of maximum entropy without any explicit regularization. Adding entropy pressure (at any tested weight) consistently hurts performance: the model is forced toward a more uniform assignment, which reduces the specialization that drives good surface accuracy.

Interestingly, the no-penalty diagnostic is slightly better than the 0.01/0.001 runs, confirming that the entropy penalty is actively harmful rather than just neutral.

### Suggested follow-ups

- **Slice utilization from a different angle**: Rather than entropy (which just measures dispersion), consider measuring the *variance of slice token norms* — if all slices converge to similar representations, the model is wasting capacity even with high marginal entropy.
- **Fix OOD Re overflow**: This pre-existing overflow (`val_ood_re/vol_loss = 18.9×10⁹`) is a separate issue that should be addressed — it's masking whether the model is actually improving on OOD Re samples.
- **This direction is closed**: Entropy reg at 0.01 and 0.001 both hurt. Unless there's a very specific reason to try 0.0001, I'd close this idea.

### CODE REVIEW

**Comment by tcapelle** [comment]
## Review — Request Changes

**Closest result to baseline in this round — val_loss only +2.1%. But doesn't beat baseline on any key metric.**

| Metric | This run | Baseline | Delta |
|--------|----------|----------|-------|
| val/loss | 2.2683 | 2.2217 | +2.1% |
| in_dist surf_p | 22.33 | 21.18 | +5.4% |
| ood_cond surf_p | 20.50 | 20.47 | +0.1% (same) |
| tandem surf_p | 41.13 | 41.23 | -0.2% (marginal) |

Your analysis is spot on: softmax already resists single-slice collapse by construction. The explicit entropy penalty may be fighting the task loss rather than complementing it.

**Next iteration — two things:**

1. **Verify the premise first.** Add a logging line in the baseline forward pass to measure actual slice entropy. If it's already near log(32)≈3.47 nats, the hypothesis is false and we should close this.

2. **If entropy is low, try a much gentler weight (0.001).** The 0.01 weight may be too aggressive, forcing uniform assignment that hurts task performance. A 10x lower weight would be a soft nudge.

Run both: first a diagnostic run logging entropy without the penalty, then if entropy is low, try 0.001.
```bash
python train.py --agent violet --wandb_name "violet/slice-entropy-diagnostic" --wandb_group slice-entropy-reg
```
